ambry-cloud/src/main/java/com/github/ambry/cloud/VcrReplicaThread.java

```diff
@@ -76,6 +94,88 @@ public VcrReplicaThread(String threadName, FindTokenHelper findTokenHelper, Clus
+    for (int i = 0; i < replicasToReplicatePerNode.size(); i++) {
+      ExchangeMetadataResponse metadata = exchangeMetadataResponseList.get(i);
+      RemoteReplicaInfo replica = replicasToReplicatePerNode.get(i);
+      if (metadata.serverErrorCode != ServerErrorCode.No_Error) {
+        // bad replica
+        continue;
+      }
+      /**
+       * Create a map of blob-id -> List<replica>
+       *   For eg.:
+       *
+       *   blob-id1 -> {replica-1, replica-2, replica-3} <sorted>
+       *   blob-id2 -> {replica-1, replica-2} <sorted>
+       *   blob-id3 -> {replica-1, replica-3} <sorted>
+       *   blob-id4 -> {replica-4, replica-5} <sorted>
+       *   blob-id3 -> {replica-4} <sorted>
+       *
+       */
+      for (MessageInfo messageInfo : metadata.getMissingStoreMessages()) {
+        String blobId = messageInfo.getStoreKey().getID();
+        List<Pair<RemoteReplicaInfo, MessageInfo>> replicaMessagePairList =
+            blobToReplicas.getOrDefault(blobId, new ArrayList<>());
+        replicaMessagePairList.add(new Pair<>(replica, messageInfo));
+        blobToReplicas.putIfAbsent(blobId, replicaMessagePairList);
+      }
+    }
+
+    for (Map.Entry<String, List<Pair<RemoteReplicaInfo, MessageInfo>>> entry : blobToReplicas.entrySet()) {
+      /**
+       * For each blob, sort the replicas in alphabetical order and just pick the first one.
+       * We can pick using any scheme, so why not alphabetical ? easy and consistent.
+       * Next, for the picked replica, collect all the blobs belonging to it.
+       */
+      List<Pair<RemoteReplicaInfo, MessageInfo>> replicaMessagePairList = entry.getValue();
+      replicaMessagePairList.sort(new LexicographicComparator());
+      Pair<RemoteReplicaInfo, MessageInfo> replicaMessagePair = replicaMessagePairList.get(0);
+      HashSet<MessageInfo> missingMessages =
+          replicaToMessages.getOrDefault(replicaMessagePair.getFirst(), new HashSet<>());
+      missingMessages.add(replicaMessagePair.getSecond());
+      replicaToMessages.putIfAbsent(replicaMessagePair.getFirst(), missingMessages);
+    }
+
+    /**
+     * Construct the exchangeMetadataResponseList
+     */
+    List<ExchangeMetadataResponse> newMetadataResponseList = new ArrayList<>();
+    for (int i = 0; i < replicasToReplicatePerNode.size(); i++) {
+      ExchangeMetadataResponse oldMetadata = exchangeMetadataResponseList.get(i);
+      RemoteReplicaInfo replica = replicasToReplicatePerNode.get(i);
+      ExchangeMetadataResponse newMetadata;
+      if (oldMetadata.serverErrorCode != ServerErrorCode.No_Error) {
+        newMetadata = oldMetadata;
+      } else if (replicaToMessages.containsKey(replica)) {
+        newMetadata = new ExchangeMetadataResponse(oldMetadata, replicaToMessages.get(replica));
+      } else {
+        newMetadata = new ExchangeMetadataResponse(oldMetadata, new HashSet<>());
+      }
+      newMetadataResponseList.add(newMetadata);
+    }
+
```

Comment:
I think this method needs more explanation and comments. Using a blob map and replica map to exchange information between multiple RemoteReplicaGroup is fine, but it needs some synchronization to clean up the maps so we don't keep all blob ids in memory.

We can add a preReplicationCycle hook to ReplicaThread and override it here to clear the blob and replica map.